pex/interpreter.py

```diff
@@ -290,32 +293,67 @@ def _execute(cls, binary, args=None, pythonpath=None, env=None, stdin_payload=No
     stdout, stderr = Executor.execute(cmd, stdin_payload=stdin_payload, env=env, **kwargs)
     return cmd, stdout, stderr
 
+  INTERP_INFO_FILE = 'INTERP-INFO'
+
   @classmethod
   def _spawn_from_binary_external(cls, binary):
-    pythonpath = third_party.expose(['pex'])
-    cmd, env = cls._create_isolated_cmd(
-      binary,
-      args=[
-        '-c',
-        dedent("""\
-        import sys
-        from pex.interpreter import PythonIdentity
-
-        sys.stdout.write(PythonIdentity.get().encode())
-        """)
-      ],
-      pythonpath=pythonpath
-    )
-    process = Executor.open_process(cmd, env=env, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    job = Job(command=cmd, process=process)
-
     def create_interpreter(stdout):
       identity = stdout.decode('utf-8').strip()
       if not identity:
         raise cls.IdentificationError('Could not establish identity of %s' % binary)
       return cls(binary, PythonIdentity.decode(identity))
 
-    return SpawnedJob.stdout(job, result_func=create_interpreter)
+    # Part of the PythonInterpreter data are environment markers that depend on the current OS
+    # release. That data can change when the OS is upgraded but (some of) the installed interpreters
+    # remain the same. As such, include the OS in the hash structure for cached interpreters.
+    os_digest = hashlib.sha1()
+    for os_identifier in platform.release(), platform.version():
+      os_digest.update(os_identifier.encode('utf-8'))
+    os_hash = os_digest.hexdigest()
```

Comment:
Nice! This might save us from awkward, hard-to-find bugs later on, I would guess.